### PR TITLE
Update simple service and client tutorial with synchronous example

### PR DESCRIPTION
**Pull request by mm318**
The current tutorial only has an example of the client using the `call_async()` API in `rclpy`. This change adds an example using the synchronous `call()` API.

### CODE REVIEW

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 1]
```diff
@@ -166,9 +169,13 @@ Add the following line between the ``'console_scripts':`` brackets:
 
   'service = py_srvcli.service_member_function:main',
 
+
 3 Write the client node
 ^^^^^^^^^^^^^^^^^^^^^^^
 
+3.1 Asynchronous client node code (makes non-blocking requests)
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]:
```suggestion
3.1 Asynchronous client node code (non-blocking requests)
```

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 2]
```diff
@@ -240,8 +245,85 @@ The only significant difference in the client’s ``main`` is the ``while`` loop
 The loop tries to check whether there is a response from the service, as long as the system is running.
 If the service has sent a response, the result will be written in a log message.
 
+3.2 Synchronous client node code (makes blocking requests)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Inside the ``dev_ws/src/py_srvcli/py_srvcli`` directory, create a new file called ``synchronous_client_member_function.py`` and paste the following code within:
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Create a new file ``synchronous_client_member_function.py`` in ``dev_ws/src/py_srvcli/py_srvcli``:

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 3]
```diff
@@ -240,8 +245,85 @@ The only significant difference in the client’s ``main`` is the ``while`` loop
...
+
+      def send_request(self):
+          self.req.a = int(sys.argv[1])
+          self.req.b = int(sys.argv[2])
+          return self.cli.call(self.req)
+
+
+  def spin(client):
+      while rclpy.ok():
+          rclpy.spin_once(client)
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Isn't there a spin function in rclpy?

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 4]
```diff
@@ -240,8 +245,85 @@ The only significant difference in the client’s ``main`` is the ``while`` loop
...
+
+
+  def main(args=None):
+      rclpy.init(args=args)
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Do we need `args` here? `main` is called once in this example, w/o any argument, so it's better to reduce complexity for the reader here and drop the arg altogether imho

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 5]
```diff
@@ -240,8 +245,85 @@ The only significant difference in the client’s ``main`` is the ``while`` loop
...
+      minimal_client.destroy_node()
+      rclpy.shutdown()
+
+
+  if __name__ == '__main__':
+      main()
+
+*Examine the code*
+
+There may be few differences between the synchronous and the asynchronous client code, but the differences are significant.
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: This sentence is not necessary, we can list the differences directly.

Compared to asynchronous client code, ``send_request`` blocks until an answer is received, and returns it.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 6]
```diff
@@ -240,8 +245,85 @@ The only significant difference in the client’s ``main`` is the ``while`` loop
...
+
+Instead of needing the query a future, ``send_request()`` directly gives back a response back.
+
+However, since ``send_request()`` is now blocking, and now both ``send_request()`` and ``spin(minimal_client)`` are blocking, one of them needs to be done in a separate thread.
+
+One may get clever and nest a synchronous ``send_request()`` in a callback, and additionally rejoice that a separate thread is not needed:
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: It is preferable to keep a technical (objective) tone in docs:

```
⚠️ It is not possible to call a service synchronously in a callback. 
etc.
```

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 7]
```diff
@@ -240,8 +245,85 @@ The only significant difference in the client’s ``main`` is the ``while`` loop
...
+  def trigger_request(msg):
+      response = minimal_client.send_request()  # this will deadlock!
+      minimal_client.get_logger().info(
+        'Result of add_two_ints: for %d + %d = %d' %
+        (minimal_client.req.a, minimal_client.req.b, response.sum))
+  subscription = minimal_client.create_subscription(String, 'trigger', trigger_request, 10)
+
+  spin(minimal_client)
+
+However, this will result in a deadlock because spinning would not preempt the callback with the ``send_request()`` call. The asynchronous ``send_request()`` is more suited to this case.
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: A better advice would be to keep callback code minimal and avoid doing expensive operations in callbacks.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 8]
```diff
@@ -240,8 +245,85 @@ The only significant difference in the client’s ``main`` is the ``while`` loop
...
+
+3.3 Add an entry point
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Add entry points ?

I guess we're adding two now so don't we need to update this section?

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 9]
```diff
@@ -254,9 +336,11 @@ The ``entry_points`` field of your ``setup.py`` file should look like this:
       'console_scripts': [
           'service = py_srvcli.service_member_function:main',
           'client = py_srvcli.client_member_function:main',
+          'synchronous_client = py_srvcli.synchronous_client_member_function:main'
       ],
   },
 
+
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Let's avoid introducing unrelated aesthetic change. Typically linters will report double empty lines, so this may trigger issues down the road. And generally whoever starts the doc get to make those minor decisions and we should respect them (follow the existing style)

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 10]
```diff
@@ -82,9 +82,13 @@ Add the same information to the ``setup.py`` file for the ``maintainer``, ``main
     description='Python client server tutorial',
     license='Apache License 2.0',
 
+
 2 Write the service node
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
+2.1 Service node code
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: All other titles start with a verb, so it'd be better to align.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst [hunk 11]
```diff
@@ -240,8 +245,85 @@ The only significant difference in the client’s ``main`` is the ``while`` loop
 The loop tries to check whether there is a response from the service, as long as the system is running.
 If the service has sent a response, the result will be written in a log message.
 
+3.2 Synchronous client node code (makes blocking requests)
```
Comment (thomas-moulard) [inline source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]:
```suggestion
3.2 Synchronous client node code
```

**Review by thomas-moulard — COMMENTED** [review]
A few issues with this change:

* it's probably easier for the reader to start with the sync version, than have the async
* the difference between sync and async is never explained (except in headers). Adding a single sentence to state the difference between the two would be useful
* Some sentences are more verbose than necessary, and do not convey a technical tone. We have very good internal guidelines for writing text, I would suggest applying the internal guidelines here too (for instance: avoiding passive, avoiding weasel words, etc.).
* The tone is not always right for developer doc. Here is a public link explaining this: https://developers.google.com/style/tone#examples_1

See notably the first example at the bottom:
"Dude! This API is totally awesome!" ➡️ "This API lets you collect data about what your users like."

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 4 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: I put this here to be consistent with the other examples. If I remove it here, I think I should remove it from the other examples as well.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 9 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: The issue was that some headings were separated by double empty lines and some by a single empty line (so I don't think there will be linting issues). I think it should be made consistent (I decided to go with the double empty lines).

Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Actually, I realized there were many more instances of single empty line separators than double empty line separators, so I'll make all the headings separated by a single empty line.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 1 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Removed the contents in the brackets altogether.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 2 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Done.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 3 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Switched over to using `rclpy.spin()` as `Thread` entry point.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 5 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Removed this line.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 6 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Done.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 7 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Done.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 8 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Done.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 10 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Done.

File: source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst (hunk 11 above)
Comment (mm318) [reply source/Tutorials/Writing-A-Simple-Py-Service-And-Client.rst]: Done.

**Comment by mm318** [comment]
Closing this draft pull request as a real one has been opened against the `ros2` organization.
